interpreter/cling/lib/Interpreter/DynamicLibraryManagerSymbol.cpp

```diff
@@ -433,29 +507,99 @@ namespace cling {
         if (!WasInserted)
           continue;
 
-        std::error_code EC;
-        for (llvm::sys::fs::directory_iterator DirIt(DirPath, EC), DirEnd;
-             DirIt != DirEnd && !EC; DirIt.increment(EC)) {
-
-          // FIXME: Use a StringRef here!
-          std::string FileName = getRealPath(DirIt->path());
-          assert(!llvm::sys::fs::is_symlink_file(FileName));
-
+        std::function<const LibraryPath*(std::string, unsigned)> HandleLib =
+            [&](std::string FileName, unsigned level) -> const LibraryPath* {
           if (ShouldPermanentlyIgnore(FileName))
-            continue;
+            return nullptr;
 
           std::string FileRealPath = llvm::sys::path::parent_path(FileName);
           FileName = llvm::sys::path::filename(FileName);
           const BasePath& BaseP = m_BasePaths.RegisterBasePath(FileRealPath);
           LibraryPath LibPath(BaseP, FileName);
-          if (m_SysLibraries.HasRegisteredLib(LibPath) ||
-              m_Libraries.HasRegisteredLib(LibPath))
-            continue;
 
+          if (auto result = m_SysLibraries.GetRegisteredLib(LibPath))
+            return &(*result);
+          if (auto result = m_Libraries.GetRegisteredLib(LibPath))
+            return &(*result);
+
+          const LibraryPath* NewLibPath;
           if (searchSystemLibraries)
-            m_SysLibraries.RegisterLib(LibPath);
+            NewLibPath = m_SysLibraries.RegisterLib(LibPath);
           else
-            m_Libraries.RegisterLib(LibPath);
+            NewLibPath = m_Libraries.RegisterLib(LibPath);
```

Comment:
Would it make sense to register the library *with* its dependencies, instead of registering it, finding it back, const-casting what we found, and then providing the dependencies?

Reply:
That is a good point, however, we should recursively, in-depth (depth-first search) accumulate the dependency list. The current algorithm is consistent to the behavior of the static linker. Moreover, the recursion in the addition of the dependencies is at scanning time, which makes the search more efficient as we do not need to recurse.

The proposed change is likely implementable but will require reorganization and performance tuning -- a task that we are not sure it is worth to avoid a minor (in our point) issue.